docs(openid-connect) update docs to match 1.5 release (#2118)
Kong/docs.konghq.com: app/_hub/kong-inc/openid-connect/1.3-x.md

3. Have rate-limiting configured for 10,000 request per hour for consumer X

With this use case you could use these parameters:

* `config.consumer_claim="sub"`
* `config.consumer_optional=true`

Default       | Required
:------------:|:-------:
`false`       | `no`

**Examples:**

* `true`
* `false`

#### config.anonymous

If consumer is not found and `config.consumer_claim` is defined,
you can use this value to fallback to an anonymous consumer. Set
this value to `id` of the consumer that you want to fallback. If
fallback fails `401` will be returned.

Default | Required
:------:|:-------:
`none`  | `no`

**Examples:**

* `"f4bc9005-7488-4fe5-9861-17a4e7cb31eb"`

#### config.run_on_preflight

A boolean value that indicates whether the plugin should run
(and try to authenticate) on `OPTIONS` preflight requests,
if set to false then `OPTIONS` requests will always be allowed.

Default | Required
:------:|:-------:
`true`  | `no`

#### config.leeway

This configuration parameter is used to adjust possible clock skew
between the `OP` and `Kong`. The `config.leeway` is taken in account
in all the time related verifications. The `leeway` is specified in

> seconds. Leeway is also used when checking access token expiration,

Reviewer: ```suggestion
seconds. Leeway is also used when checking access token expiration;
```

i.e. when to refresh the access token if refresh token is available.

#### config.verify_parameters

> For debugging purposes you can turn off parameter verification. Sometimes

Reviewer: ```suggestion
For debugging purposes, you can turn off parameter verification. Sometimes
```

providers announce erroneous information in their discovery documents

> (e.g. they don't announce features that they actually do support). This

Reviewer: ```suggestion
(e.g., they don't announce features that they actually do support). This
```

> parameter turns all the discovery based parameter validations off. It can

Reviewer: ```suggestion
parameter turns all the discovery-based parameter validations off. It can
```

> be safely turned off. Please always check Kong error logs, as the plugins

Reviewer: ```suggestion
be safely turned off. Always check Kong error logs, as the plugins
```

in this suite usually give very clear information of what went wrong.

#### config.verify_nonce

> For debugging purposes you can turn off nonce verification. Please always

Reviewer: ```suggestion
For debugging purposes, you can turn off nonce verification. Always
```

check Kong error logs, as the plugins in this suite usually give very clear
information of what went wrong. Turning off the verification of nonce is not

> a good idea in production as it is a security feature, If your identity provider

Reviewer: ```suggestion
a good idea in production as it is a security feature. If your identity provider
```

doesn't support it, then it might be reasonable to try turning this off.

#### config.verify_claims

You can turn off all the claims verification with this option. Generally
it is not recommended, but you can use this for debugging purposes.
A signature will still be verified.

> Additional claims verification (e.g. `config.scopes_required`) will still

Reviewer: ```suggestion
Additional claims verification (e.g., `config.scopes_required`) will still
```

be checked if defined.

#### config.verify_signature

You can turn off all the signature verification with this option. Generally
it is not recommended, but you can use this for debugging purposes.
Claims will still be verified.

#### config.ignore_signature

You can turn off access token signature verification on specified flows or
grant types. This feature was added because Microsoft Azure AD gives JWT
access tokens that point to a JWT with `kid` that Microsoft has not published
to public, thus it is impossible to verify signature of such tokens.

You can use following values:

* `password`
* `client_credentials`
* `authorization_code`
* `refresh_token`
* `session`

The `bearer` is not an option for security purposes. If you really want to
skip bearer signature verification, then turn `config.verify_signature` off,
but that practically means that any token can be sent.

#### config.cache_ttl

> The default cache expiry time in seconds. In many cases the expiry is stored

Reviewer: ```suggestion
The default cache expiry time in seconds. In many cases, the expiry is stored
```

in tokens (`exp` claim) or token endpoint response (`expires_in` claim),
and those are used when available. When they are not available, the
`config.cache_ttl` gets used.

#### config.cache_ttl_max

> Maximum time in seconds to store items (e.g. tokens) in a cache.

Reviewer: ```suggestion
Maximum time to live in seconds to store items (e.g., tokens) in a cache.
```

#### config.cache_ttl_min

> Minimum time in seconds to store items (e.g. tokens) in a cache.

Reviewer: ```suggestion
Minimum time to live in seconds to store items (e.g. tokens) in a cache.
```

#### config.cache_ttl_neg

> Specifies time in seconds of how long to cache negative results.

Reviewer: ```suggestion
Specifies time to livein seconds of how long to cache negative results.
```

#### config.cache_ttl_resurrect

When specified, the cache instance will attempt to resurrect stale values when the L3 callback returns nil, err (soft errors).

#### config.cache_tokens

You can turn off caching of token endpoint requests with this parameter.

#### config.cache_introspection

You can turn off caching of introspection requests with this parameter.

#### config.cache_token_exchange

You can turn off caching of token exchange requests with this parameter.

#### config.cache_user_info

> You can turn off caching of user info request with this parameter.

Reviewer: ```suggestion
You can turn off caching of user info requests with this parameter.
```

#### config.search_user_info

Enables or disables searching for the `consumer_claim`, `credential_claim` and
`authenticated_groups` in [UserInfo response data](https://openid.net/specs/openid-connect-core-1_0.html#UserInfo)
if it was not found in the ID token, access token, or introspection response.

Default | Required
:------:|:-------:
`false` | `no`

**Examples:**

* `true`
* `false`

#### config.hide_credentials

An optional boolean value telling the plugin to hide the credential to

> the upstream API server. It will be removed by Kong before proxying

Reviewer: ```suggestion
the Upstream API server. It will be removed by Kong before proxying
```

the request.

#### config.http_version

Specifies the HTTP version that is used for communicating with the

> OP. It can either be `1.1` or `1.0`.

Reviewer: ```suggestion
OP (Open ID Connect Provider). It can either be `1.1` or `1.0`.
```

Default | Required
:------:|:-------:
`1.1`   | `no`

**Examples:**

* `1.0`
* `1.1`

#### config.http_proxy

Specifies the HTTP proxy to be used for communicating with the IdP.

**Examples:**

* `http://10.0.0.5:8080`

#### config.http_proxy_authorization

A value for `Proxy-Authorization` header to be used with `config.http_proxy`.

**Examples:**

* `Basic ZGVtbzp0ZXN0`

#### config.https_proxy

Specifies the HTTPS proxy to be used for communicating with the IdP.

**Examples:**

* `http://10.0.0.5:8080`

#### config.https_proxy_authorization

A value for `Proxy-Authorization` header to be used with `config.https_proxy`.

**Examples:**

* `Basic ZGVtbzp0ZXN0`

#### config.no_proxy

A comma-separated list of hosts that should not be proxied.

**Examples:**

* `example.org,example.com`

#### config.keepalive

> Specifies whether or not the plugin should keep-alive connections

Reviewer: ```suggestion
Specifies whether the plugin should keep-alive connections
```

with the IdP endpoints.

Default | Required
:------:|:-------:
`true` | `no`

**Examples:**

* `true`
* `false`

#### config.ssl_verify

> Specifies whether or not the plugin should verify the certificates

Reviewer: ```suggestion
Specifies whether the plugin should verify the certificates
```

used in SSL/TLS communication.

Default | Required
:------:|:-------:
`false` | `no`

**Examples:**

* `true`
* `false`

#### config.timeout

 Specified the timeouts for HTTP requests used in non-blocking communications
 between `Kong` and `OP`, and between `Kong` and the `Client`. This parameter
 is specified in milliseconds.

 Default | Required
 :------:|:-------:
 `10000`  | `no`

 **Examples:**

 * `10000`
 * `30000`
 * `5000`

### Authorization

> `openid-connect` plugin supports different authentication / authorization

Reviewer: ```suggestion
`openid-connect` plugin supports different authentication/authorization
```

Reviewer: @bungle this one needs inc. in both 1.3 and 1.5

methods in one plugin. The plugin figures out the method used for

> authentication by looking the request headers or parameters (and also

Reviewer: ```suggestion
authentication by looking at the request headers or parameters (and also
```

plugin configuration parameter `config.auth_methods`).

#### Authorization Code Grant

> Authorization Code grant is default (if enabled) that Kong tries when there is

Reviewer: ```suggestion
Authorization Code grant (if enabled) is the default grant type that Kong tries when there are
```

> no other credentials supplied. With authorization code flow Kong sends a

Reviewer: ```suggestion
no other credentials supplied. With authorization code flow, Kong sends a
```

HTTP redirect response to the client, and initiates the authorization code flow.
After the client authenticates on identity provider, the provider sends the
client back to Kong where Kong validates the parameters and then goes on with

> exchanging the authorization code with the tokens by calling token endpoint

Reviewer: ```suggestion
exchanging the authorization code with the tokens by calling the token endpoint
```

of the identity provider.

#### Resource Owner Password Credentials Grant

The plugin looks for `Authorization: Basic <username>:<password>` header to
determine the username and password. `<username>:<password>` pair can (and should)

> be base64 encoded (HTTP Basic Authentication). Alternatively you can send

Reviewer: ```suggestion
be base64encoded (HTTP Basic Authentication). Alternatively, you can send
```

`username` and `password` in url (`GET`) or body (`POST`) arguments using
`username` and `password` fields.

#### Client Credentials Grant

The plugin looks for `Authorization: Basic <client_id>:<client_secret>` header to
determine the username and password. `<client_id>:<client_secret>` pair can (and should)

> be base64 encoded (HTTP Basic Authentication). Alternatively you can send

Reviewer: ```suggestion
be base64-encoded (HTTP Basic Authentication). Alternatively, you can send
```

Reviewer: @bungle I missed the hyphen here both 1.3 and 1.5

Reviewer: done in commit 9f48d19

`client_id` and `client_secret` in url (`GET`) or body (`POST`) arguments using
`client_id` and `client_secret` fields.

#### JWT Bearer Token

> When the client has an JWT access token available. E.g. it could have retrieved one

Reviewer: Incomplete first sentence.

directly from identity provider by using implicit flow, or it could have received it
by other means. That token an then be used as a bearer token that is presented in
an authorization header (`Authorization: Bearer <token>`).  The JWT token (or JWS
in this case) will be verified for signature and standard claims before it gets
accepted.

#### Opaque Bearer Token

Similar to JWT bearer tokens, but this time Kong will figure out through introspection
whether or not to trust this token. Also the opaque access tokens are presented in an
authorization header (`Authorization: Bearer <token>`).

#### Kong OAuth 2.0 Authentication Plugin issued Opaque Access Tokens

The plugin will also accept tokens that are issued by Kong OAuth 2.0 Authentication Plugin,
and it can be used to verify them. It duplicates some verification code from Kong OAuth 2.0
plugin, but this plugin does not provide any identity provider functionality as the Kong OAuth

> 2.0 plugin does. Similar to JWT Bearer tokens and opaque bearer tokens, the Kong issued opaque

Reviewer: ```suggestion
2.0 plugin does. Similar to JWT Bearer tokens and opaque bearer tokens, the Kong-issued opaque
```

tokens are presented in an authorization header (`Authorization: Bearer <token>`).

#### Session Cookie

When the client has authenticated with one of the supported authentication methods mentioned

> above, this plugin can also (optionally) send a HTTP Only session cookie to the client. This

Reviewer: ```suggestion
above, this plugin can also (optionally) send an HTTP Only session cookie to the client. This
```

> cookie can be used to authenticate further requests This works especially nicely with

Reviewer: ```suggestion
cookie can be used to authenticate further requests This works especially well with
```

Authorization Code Grant that is usually used with interactive browser sessions. Session cookie

> is presented to this plugin by setting the cookie header (e.g. `Cookie: session=…`).

Reviewer: ```suggestion
is presented to this plugin by setting the cookie header (e.g., `Cookie: session=…`).
```

#### Refresh Token (or Offline Token)

If the client provides a refresh token or offline token to this plugin, the plugin can
attempt to fetch tokens from the token endpoint using `refresh_token` grant. It is necessary
to configure the parameter name where the refresh token is supplied with
`config.refresh_token_param_name` and `config.refresh_token_param_type` configuration parameters.

### Usage

Here are the step by step usage instructions. The examples are executed
using [HTTPie](https://httpie.org/). Before proceeding, please **NOTE**
that this usage example is for testing purposes and that you should
not send confidential information to `httpbin.org` that is used here

> for illustrative purposes.

Reviewer: ```suggestion
for demonstration purposes.
```

> #### 1. Creating a Service and Route

Reviewer: ```suggestion
#### 1. Create a Service and Route
```

To enable the OIDC Plugin for the Service, execute the following
command, but note that SSL verification is disabled here for
testing purposes only:

```bash
http post :8001/services/httpbin/plugins \
  name=openid-connect                             \
  config.issuer=<ISSUER>                          \
  config.client_id=<CLIENT_ID>                    \
  config.client_secret=<CLIENT_SECRET>            \
  config.redirect_uri=<REDIRECT_URI>              \
  config.ssl_verify=false -v
```

On successful `HTTP 200 OK`, the response will be similar to:

```json
{
  "config": {
    "audience_claim": [
      "aud"
    ],
    "client_id": [
      "kong"
    ],
    "client_secret": [
      "b8068d7d-d7bf-4b23-8724-881ee49bdbfd"
    ],
    "consumer_by": [
      "username",
      "custom_id"
    ],
    "introspect_jwt_tokens": false,
    "introspection_hint": "access_token",
    "issuer": "http://ip10-0-0-3-bec2g5kmeb6ge43qmuc0-8080.direct.konglabs-s3.simplru.com/auth/realms/master",
    "created_at": 1536700540000,
    "enabled": true,
    "id": "6836ba3c-12e0-4f2e-bdfd-88134eaa1786",
    "name": "openid-connect",
    "service_id": "903978c9-2472-4a04-aff5-d9ba13821e64"
  }
}
```

> #### 4. Verify that Authorization is Now Required

Reviewer: ```suggestion
#### 4. Verify that Authorization is now required
```

Attempt the following command without authorization:

```bash
$ http get :8000/anything
```

It will result in:

```
HTTP 301 Moved Temporarily
```

Now, at this point you could try to open the page using a browser and
see if you can go through the authorization code flow, and after that

> get a reply from `httpbin.org`. Please check that your redirect URI parameter is

Reviewer: ```suggestion
get a reply from `httpbin.org`. Ensure that your redirect URI parameter is
```

correctly set, since the identity provider should redirect the

> user back to this URI once authentication is successful. (The URI can be the same

Reviewer: ```suggestion
user back to this URI once authentication is successful. The URI can be the same
```

> Service or a different one).

Reviewer: ```suggestion
Service or a different one.
```
sentence can standalone w/out parens
